### ISSUE TITLE
Install issues

### ISSUE THREAD

**Issue by j1elo** [sernst/cauldron#51]
There are multiple issues installing on Linux Mint 18.3, which is basically an Ubuntu 16.04 ("Xenial") system.

By following the instructions on the [home page](http://www.unnotebook.com/install/):

1. Install Python package, making sure to use Python 3 tools:

```
$ sudo apt-get install python3-pip
$ sudo pip3 install --upgrade cauldron-notebook
```
This installs `/usr/local/bin/cauldron`, which gives problems later...
```
$ sudo updatedb && locate -i "bin/cauldron"
/usr/local/bin/cauldron
```
Apart from that, the Cauldron's Python package is correctly installed, together with all other PIP packages:
```
$ sudo updatedb && locate -i "cauldron" | grep __init__.py | head -n1
/usr/local/lib/python3.5/dist-packages/cauldron/__init__.py
```

2. Install application package:

```
$ sudo dpkg -i cauldron_3.221.532_amd64.deb
# Install finishes without errors
```
This installs `/usr/bin/cauldron`:
```
$ sudo updatedb && locate -i "bin/cauldron"
/usr/bin/cauldron
/usr/local/bin/cauldron
```

3. Try to run:

Calling `cauldron` will run `/usr/local/bin/cauldron` due to how binaries are searched at execution time (user local binaries are preferred over system default ones). Also, it breaks:
```
$ cauldron 
Traceback (most recent call last):
  File "/usr/local/bin/cauldron", line 7, in <module>
    from cauldron.invoke import run
  File "/usr/local/lib/python3.5/dist-packages/cauldron/__init__.py", line 1, in <module>
    from cauldron import session as _session
  File "/usr/local/lib/python3.5/dist-packages/cauldron/session/__init__.py", line 6, in <module>
    from cauldron.session.exposed import ExposedProject
  File "/usr/local/lib/python3.5/dist-packages/cauldron/session/exposed.py", line 9, in <module>
    from cauldron.render import stack as render_stack
  File "/usr/local/lib/python3.5/dist-packages/cauldron/render/stack.py", line 5, in <module>
    from cauldron.session import projects
  File "/usr/local/lib/python3.5/dist-packages/cauldron/session/projects/__init__.py", line 1, in <module>
    from cauldron.session.projects.project import DEFAULT_SCHEME  # noqa
  File "/usr/local/lib/python3.5/dist-packages/cauldron/session/projects/project.py", line 14, in <module>
    from cauldron.session.projects import steps
  File "/usr/local/lib/python3.5/dist-packages/cauldron/session/projects/steps.py", line 12, in <module>
    from cauldron.session.report import Report
  File "/usr/local/lib/python3.5/dist-packages/cauldron/session/report.py", line 11, in <module>
    class Report(object):
  File "/usr/local/lib/python3.5/dist-packages/cauldron/session/report.py", line 45, in Report
    def project(self) -> typing.Union['projects.Project', None]:
  File "/usr/lib/python3.5/typing.py", line 552, in __getitem__
    dict(self.__dict__), parameters, _root=True)
  File "/usr/lib/python3.5/typing.py", line 512, in __new__
    for t2 in all_params - {t1} if not isinstance(t2, TypeVar)):
  File "/usr/lib/python3.5/typing.py", line 512, in <genexpr>
    for t2 in all_params - {t1} if not isinstance(t2, TypeVar)):
  File "/usr/lib/python3.5/typing.py", line 190, in __subclasscheck__
    self._eval_type(globalns, localns)
  File "/usr/lib/python3.5/typing.py", line 177, in _eval_type
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
AttributeError: module 'cauldron.session.projects' has no attribute 'Project'
```

Giving the full path does actually run the correct program, but it doesn't find the `cauldron` Python package:
```
$ /usr/bin/cauldron
```

**First screen**:
```
To Start Cauldron...
Enter your Python 3.5+ executable path
or a remote Cauldron connection URL
> /usr/bin/python3
```

**Second screen**:
```
Missing Cauldron
Unable to import cauldron python package. Has it been installed?
```

**Comment by sernst** [comment]
Hi, @j1elo. Yeah, the conflicting application names is an issue. I'll look into renaming the GUI installer to install `cauldron-ui` instead of  `cauldron` in the deb package.

The `AttributeException` that you encountered is due to a bug:

https://bugs.python.org/issue26477

related to forward referencing in the typing module:

https://www.python.org/dev/peps/pep-0484/#forward-references

that was fixed in 3.5.3 (current version is 3.5.6). Unfortunately,  Ubuntu 16.04 appears to still install 3.5.2 by default. My recommendation would be to use an alternative method to install a more 
recent version of Python.

One way would be to use the deadsnakes PPA, but that would mean updating to Python 3.6 because the PPA does not include Python 3.5 packages for Xenial to avoid conflicts with the system installation. 
This could be done with the commands:

```bash
$ apt-get install software-properties-common
$ add-apt-repository ppa:deadsnakes/ppa
$ apt-get update
$ apt-get install python3.6
```

The other possibility would be to install a more recent version of Python 3.5 from source as a user-installed package instead of a system installed one to prevent conflicts with the system-required package.

I'll update the installation information in the site to clarify path versions of 3.5 to prevent this confusion in the future.

**Comment by j1elo** [comment]
Hi,

> The `AttributeException` that you encountered is due to a bug: ...

Is this the reason why the GUI doesn't seem to find the Python package?
```
Missing Cauldron
Unable to import cauldron python package. Has it been installed?
```

If at all possible, I'd suggest not requiring your users having to fiddle with their systems, and instead adapting your software with feature flags or even avoiding unsupported features altogether. It's unfortunate that Ubuntu 16.04 didn't patch up their version to fix this bug (they do for security bugs but not for feature issues) but there are probably still lots of people using it. I'm not talking for myself here, I know pretty well what I'm doing. Just a suggestion to make the project more accessible.

Regarding the conflicting binary names, I'd suggest using `-gui` instead of `-ui`, as well, it *is* a graphical interface after all... also seems that `program-gui` is much more popular naming than `program-ui`:

```
$ apt-file search -x 'bin/.*-gui$' | uniq -u | wc -l
54
$ apt-file search -x 'bin/.*-ui$' | uniq -u | wc -l
13
```

**Comment by sernst** [comment]
When the GUI tries to connect to python it first checks to make sure that
the cauldron package can be imported by basically running:

`python -c "import cauldron; print(cauldron.__version__)"`

Right now I'm checking the returncode from that command execution and showing:

```
Missing Cauldron
Unable to import cauldron python package. Has it been installed?
```

for any non-zero value. I'm not parsing stderr/stdout to determine
the actual cause of the import error. So in your case the
`AttributeException` is making it appear as though the cauldron cannot
be found. I'll create an issue to look into more granualar errors for
failures from that test.

Yes, it is unfortunate that Ubuntu doesn't keep up with patch releases. 
There have been security fixes in 3.5.x since 3.5.2 was released but I
guess Ubuntu doesn't consider those to be critical:

https://docs.python.org/3.5/whatsnew/changelog.html

Yep, agreed. A `-gui` suffix makes sense.

**Comment by sernst** [comment]
This has been resolved in Cauldron 1.0.0, which has switched from a desktop-based UI to a web-based one.
